autorest/codegen/serializers/utils.py

```diff
@@ -102,13 +102,36 @@ def method_signature_and_response_type_annotation_template(
         return f"{method_signature} -> {response_type_annotation}:"
     return f"{method_signature}:\n    # type: (...) -> {response_type_annotation}"
 
-def pop_kwargs_from_signature(kwargs_to_pop: List[Parameter]) -> List[str]:
+def pop_kwargs_from_signature(
+    kwargs_to_pop: List[Parameter],
+    check_kwarg_dict: bool,
+    pop_headers_kwarg: bool,
+    pop_params_kwarg: bool,
+    check_pop_type: bool = False,
+) -> List[str]:
     retval = []
+    def append_pop_kwarg(key: str) -> None:
+        if check_pop_type:
+            retval.append(f'_{key} = kwargs.pop("{key}", {{}}) or {{}}')
+            retval.append(f'if isinstance(_{key}, dict):')
+            retval.append(f'    _{key} = case_insensitive_dict(_{key})')
```

Comment:
I think because `headers` and `params` are reserved keywords that refer explicitly to headers and params dicts, we don't need to do checks for if it's a dict, can we just always convert `headers` and `params` to case insensitive?

Additionally, we might one day want to fully accept lists of tuples etc instead of dict, so it would fail this check

Reply:
yes, a little ugly and may be a trap in future. Updated to always use case_insensitive_dict in request builder